feat: Make it easier to specify target architecture (#869)
rp-rs/rp-hal: rp235x-hal-examples/README.md

<!-- GETTING STARTED -->
## Getting Started

To build the examples, first grab a copy of the source code:

```console
$ git clone https://github.com/rp-rs/rp-hal.git
```

> Then use `rustup` to grab the Rust Standard Library for the appropriate targets. the RP2350
> has two possible targets, thumbv8m.main-none-eabihf for the ARM cpu. And riscv32imac-unknown-none-elf
> for the RISC-V cpu.

Reviewer: ```suggestion
Then use `rustup` to grab the Rust Standard Library for the appropriate targets. 
RP2350 has two possible targets: thumbv8m.main-none-eabihf for the ARM cpu, and riscv32imac-unknown-none-elf for the RISC-V cpu.
```

```console
$ rustup target add thumbv8m.main-none-eabihf
$ rustup target add riscv32imac-unknown-none-elf
```

**Note: all examples assume the current directory is <repo root>/rp235x-hal-examples.**
```
cd rp235x-hal-examples
```

The most basic method is to use `cargo build` with the `--bin` flag to specify the example you want to
build. For example, to build the `blinky` example:
```console
$ cargo build --bin blinky
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
   Compiling syn v1.0.109
...
   Compiling pio-parser v0.2.2

>    Compiling rp235x-hal v0.2.0 (/home/wink/prgs/rpi-pico/myrepos/rp-hal-clone/rp235x-hal)

Reviewer: ```suggestion
   Compiling rp235x-hal v0.2.0
```

   Compiling pio-proc v0.2.2
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 14.97s
```

This builds the default target, which is the for the ARM at `./target/thumbv8m.main-none-eabihf/debug/blinky`:
```console
$ file ./target/thumbv8m.main-none-eabihf/debug/blinky
./target/thumbv8m.main-none-eabihf/debug/blinky: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, with debug_info, not stripped
```

If you want to build the RISC-V you can specify the target directly to override the default:
```console
$ cargo build --target=riscv32imac-unknown-none-elf --bin blinky
   Compiling nb v1.1.0
   Compiling byteorder v1.5.0
   Compiling stable_deref_trait v1.2.0
..
   Compiling futures v0.3.31
   Compiling frunk v0.4.3

>    Compiling rp235x-hal v0.2.0 (/home/wink/prgs/rpi-pico/myrepos/rp-hal-clone/rp235x-hal)

Reviewer: ```suggestion
   Compiling rp235x-hal v0.2.0
```

    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.23s```
```

> And we see the RISC-V at `./target/riscv32imac-unknown-none-elf/debug/blinky`:

Reviewer: ```suggestion
And we can see that a RISC-V elf file is now present at `./target/riscv32imac-unknown-none-elf/debug/blinky`:
```

```console
$ file ./target/riscv32imac-unknown-none-elf/debug/blinky
./target/riscv32imac-unknown-none-elf/debug/blinky: ELF 32-bit LSB executable, UCB RISC-V, RVC, soft-float ABI, version 1 (SYSV), statically linked, with debug_info, not stripped
```

> You can also specify the ARM target directly just pass

Reviewer: ```suggestion
You can also specify the ARM target directly by using
```

`--target thumbv8m.main-none-eabihf` instead of `--target riscv32imac-unknown-none-elf`.

You can also easily build, flash and start the application on the RP235x

> by using the `cargo run` using command:

Reviewer: ```suggestion
by using the `cargo run` with one of the following commands:
```

```console
$ cargo run --target thumbv8m.main-none-eabihf --bin blinky
$ cargo run --target riscv32imac-unknown-none-elf --bin blinky
```

You can also specify a release build by passing `--release` to the `cargo build`
or `cargo run` commands. This will build the example with optimizations enabled:
```console
$ cargo run --target thumbv8m.main-none-eabihf --release --bin blinky
$ cargo run --target riscv32imac-unknown-none-elf --release --bin blinky
```

The above work well but the commands are somewhat verbose. To make building and running

> more succinct we have added some aliases in .carog/config.toml:

Reviewer: ```suggestion
more succinct we have added some aliases in .cargo/config.toml:
```

> ```toml

Reviewer 2: Link to the file rather than copy-pasting the contents.

Author: OK

# Add aliases for building and running for the ARM and RISC-V targets.
[alias]

# Build arm or riscv using defaults or other options
bld-arm = "build --target=thumbv8m.main-none-eabihf"
bld-riscv = "build --target=riscv32imac-unknown-none-elf"

# Run arm or riscv using defaults or other options
run-arm = "run --target=thumbv8m.main-none-eabihf"
run-riscv = "run --target=riscv32imac-unknown-none-elf"

# Build dev and release profiles for arm
bda = "bld-arm"
bra = "bld-arm --release"

# Build dev and release for profiles risc-v
bdr = "bld-riscv"
brr = "bld-riscv --release"

# Run dev and release for profiles arm
rda = "run-arm"
rra = "run-arm --release"

# Run dev and release for profiles risc-v
rdr = "run-riscv"
rrr = "run-riscv --release"
```

> Using the aliases the commands are much shorter the first is for
> running on ARM in rlease mode and the second is for running on RISC-V in release mode:

Reviewer: ```suggestion
When using these aliases your build/run commands are much shorter.
Lets run the blinky example again, using `cargo rra` for running on ARM in release mode and `cargo rrr` for running on RISC-V in release mode:
```

```console
$ cargo rra --bin blinky
$ cargo rrr --bin blinky
```

These are exactly the same as:
```console
$ cargo run --target thumbv8m.main-none-eabihf --release --bin blinky
$ cargo run --target riscv32imac-unknown-none-elf --release --bin blinky
```

> Here is the output of running in release mode the `blinky` example on ARM
> after cleaning the build **Note:** have the pico 2 in BOOTSEL mode:

Reviewer: ```suggestion
Here is the output of running the `blinky` example in release mode on ARM
after cleaning the build **Note:** have the pico 2 in BOOTSEL mode before running this command:
```

```console
$ cargo clean
     Removed 3565 files, 1.4GiB total
$ cargo rra --bin blinky
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
..

>    Compiling rp235x-hal v0.2.0 (/home/wink/prgs/rpi-pico/myrepos/rp-hal/rp235x-hal)

Reviewer: ```suggestion
   Compiling rp235x-hal v0.2.0
```

<!-- GETTING STARTED -->
## Getting Started

To build the examples, first grab a copy of the source code:

```console
$ git clone https://github.com/rp-rs/rp-hal.git
```

Then use `rustup` to grab the Rust Standard Library for the appropriate targets.

> RP2350 has two possible targets: thumbv8m.main-none-eabihf for the ARM cpu, and

Reviewer 2: Target should be in `backticks` and CPU should be in capitals

Reviewer 2: Same for line below

riscv32imac-unknown-none-elf for the RISC-V cpu.
```console
$ rustup target add thumbv8m.main-none-eabihf
$ rustup target add riscv32imac-unknown-none-elf
```

> **Note: all examples assume the current directory is <repo root>/rp235x-hal-examples.**

Reviewer 2: Path should be in `backticks` because otherwise it looks like HTML

```
cd rp235x-hal-examples
```

The most basic method is to use `cargo build` with the `--bin` flag to specify the example you want to
build. For example, to build the `blinky` example:
```console
$ cargo build --bin blinky
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
   Compiling syn v1.0.109
...
   Compiling pio-parser v0.2.2
   Compiling rp235x-hal v0.2.0
   Compiling pio-proc v0.2.2
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 14.97s
```

> This builds the default target, which is the for the ARM and the elf file

Reviewer 2: Typo: "which is the for the"?

Raspberry Pi call it "Arm mode" and "RISC-V mode". We should use those terms.

Author: I've tried to change all occurrences and also changed "elf" to "ELF" in the text.

is located at `./target/thumbv8m.main-none-eabihf/debug/blinky`:
```console
$ file ./target/thumbv8m.main-none-eabihf/debug/blinky
./target/thumbv8m.main-none-eabihf/debug/blinky: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, with debug_info, not stripped
```

If you want to build the RISC-V you specify the target directly to override the default:

> ```console

Reviewer 2: Need a blank line before each console block.

$ cargo run --target riscv32imac-unknown-none-elf --release --bin blinky
```

For the ARM target all of the examples are built if no `--bin` is specified:
```console
$ cargo clean
     Removed 1488 files, 398.6MiB total
$ cargo build --target thumbv8m.main-none-eabihf
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
..
   Compiling rp235x-hal v0.2.0
   Compiling pio-proc v0.2.2
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 16.08s
$ find target/thumbv8m.main-none-eabihf/debug/ -maxdepth 1 -type f -executable | sort
target/thumbv8m.main-none-eabihf/debug/adc
target/thumbv8m.main-none-eabihf/debug/adc_fifo_dma
..
target/thumbv8m.main-none-eabihf/debug/blinky
..
target/thumbv8m.main-none-eabihf/debug/vector_table
target/thumbv8m.main-none-eabihf/debug/watchdog
```

For the RISC-V it is currently possible to build only *some* of the examples. See
[`riscv_examples.txt`](./riscv_examples.txt) for a list of known working examples.
The missing ones probably rely on interrupts, or some other thing we
haven't ported to work in RISC-V mode yet.

Here is an example using the `blinky` example in RISC-V mode. We'll build and

> run it first in non-release emulating the developement cycle **Note:** be sure

Reviewer 2: Typo: development

And I think this should talk about compiling using the default debug profile, not running "in non-release"

Author: I've used "dev profile" as that seems the correct term from [here](https://doc.rust-lang.org/cargo/reference/profiles.html)

RP2350 has two possible targets: `thumbv8m.main-none-eabihf` for the Arm mode, and
`riscv32imac-unknown-none-elf` for the RISC-V mode.

```console
$ rustup target add thumbv8m.main-none-eabihf
$ rustup target add riscv32imac-unknown-none-elf
```

**Note: all examples assume the current directory is `<repo root>/rp235x-hal-examples`.**
```
cd rp235x-hal-examples
```

The most basic method is to use `cargo build` with the `--bin` flag to specify the example you want to
build. For example, to build the `blinky` example:

```console
$ cargo build --bin blinky
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
   Compiling syn v1.0.109
...
   Compiling pio-parser v0.2.2
   Compiling rp235x-hal v0.2.0
   Compiling pio-proc v0.2.2
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 14.97s
```

This builds the default target, which is Arm mode and the ELF file
is located at `./target/thumbv8m.main-none-eabihf/debug/blinky`:

```console
$ file ./target/thumbv8m.main-none-eabihf/debug/blinky
./target/thumbv8m.main-none-eabihf/debug/blinky: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, with debug_info, not stripped
```

> If you want to build the RISC-V mode you specify the target directly to override the default:

Reviewer 2: ```suggestion
If you want to build a binary that runs in RISC-V mode, then you must specify the RISC-V target to override the default:
```

```console
$ cargo build --target=riscv32imac-unknown-none-elf --bin blinky
   Compiling nb v1.1.0
   Compiling byteorder v1.5.0
   Compiling stable_deref_trait v1.2.0
..
   Compiling futures v0.3.31
   Compiling frunk v0.4.3
   Compiling rp235x-hal v0.2.0
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.23s```
```

And we see that the RISC-V mode ELF file is now present at `./target/riscv32imac-unknown-none-elf/debug/blinky`:

```console
$ file ./target/riscv32imac-unknown-none-elf/debug/blinky
./target/riscv32imac-unknown-none-elf/debug/blinky: ELF 32-bit LSB executable, UCB RISC-V, RVC, soft-float ABI, version 1 (SYSV), statically linked, with debug_info, not stripped
```

You can also specify the Arm mode target directly by using
`--target thumbv8m.main-none-eabihf` instead of `--target riscv32imac-unknown-none-elf`.

To build, flash and start the application on the RP235x

> you use `cargo run` with one of the following commands:

Reviewer 2: ```suggestion
you use `cargo run` with one of the following commands. Note: be sure the RP235x is in BOOTSEL mode before using the `run` command because we use `picotool` to flash and run the binary:
```

```

For the Arm mode target all of the examples are built if no `--bin` is specified:

```console
$ cargo clean
     Removed 1488 files, 398.6MiB total
$ cargo build --target thumbv8m.main-none-eabihf
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
..
   Compiling rp235x-hal v0.2.0
   Compiling pio-proc v0.2.2
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 16.08s
$ find target/thumbv8m.main-none-eabihf/debug/ -maxdepth 1 -type f -executable | sort
target/thumbv8m.main-none-eabihf/debug/adc
target/thumbv8m.main-none-eabihf/debug/adc_fifo_dma
..
target/thumbv8m.main-none-eabihf/debug/blinky
..
target/thumbv8m.main-none-eabihf/debug/vector_table
target/thumbv8m.main-none-eabihf/debug/watchdog
```

For the RISC-V mode it is currently possible to build only *some* of the examples. See
[`riscv_examples.txt`](./riscv_examples.txt) for a list of known working examples.
The missing ones probably rely on interrupts, or some other thing we
haven't ported to work in RISC-V mode yet.

Here is an example using the `blinky` example in RISC-V mode. We'll build and
run it first using the default dev profile, emulating the development cycle **Note:** be sure

> the pico 2 is in BOOTSEL mode before the `run` command:

Reviewer 2: ```suggestion
the RP235x is in BOOTSEL mode before using the `run` command because we use `picotool` to flash and run the binary:
```

The device was rebooted to start the application.
```

At this point the development version is running on the RP2350 and the LED is blinking.
When we look at `blinky` using `file` we see we have an RISC-V ELF file and using `ls`

> we see it is quite large, 2.4M:

Reviewer 2: The size of an ELF is mostly irrelevant. What matters are the size of the output sections (aka segments) reported by the `rust-size` command that comes with `cargo-binutils`. So we could show that here.

Or we could defer to other reference material on this subject and not go into file size.

Author: I've dropped the file size, I did install cargo-binutils and used rust-size, interesting that text is slightly larger on ARM vs RISC-V at 8692 vs 7468, but bss is 4 vs 0.5M (0x80000) in release profile:
```
$ rust-size target/thumbv8m.main-none-eabihf/release/blinky
   text	   data	    bss	    dec	    hex	filename
   8692	      0	      4	   8696	   21f8	target/thumbv8m.main-none-eabihf/release/blinky

$ rust-size target/riscv32imac-unknown-none-elf/release/blinky
   text	   data	    bss	    dec	    hex	filename
   7468	      0	 524288	 531756	  81d2c	target/riscv32imac-unknown-none-elf/release/blinky
```

The device was rebooted to start the application.
```

At this point the development version is running on the RP2350 and the LED is blinking.
When we look at `blinky` using `file` we see we have generated a RISC-V mode ELF file:

```console
$ file ./target/riscv32imac-unknown-none-elf/debug/blinky
./target/riscv32imac-unknown-none-elf/debug/blinky: ELF 32-bit LSB executable, UCB RISC-V, RVC, soft-float ABI, version 1 (SYSV), statically linked, with debug_info, not stripped
```

Next we'll build and run it using the release profile for "final" testing,

> again be sure the pico 2 is in BOOTSEL mode:

Reviewer 2: ```suggestion
again be sure the RP235x is in BOOTSEL mode:
```

```console
$ cargo run --bin blinky --target=riscv32imac-unknown-none-elf --release
   Compiling proc-macro2 v1.0.89
   Compiling unicode-ident v1.0.13
..
   Compiling pio-parser v0.2.2
   Compiling pio-proc v0.2.2
    Finished `release` profile [optimized] target(s) in 17.05s
     Running `picotool load -u -v -x -t elf target/riscv32imac-unknown-none-elf/release/blinky`
Family id 'rp2350-riscv' can be downloaded in absolute space:
  00000000->02000000
Loading into Flash: [==============================]  100%
Verifying Flash:    [==============================]  100%
  OK

The device was rebooted to start the application.
```

> The LED should be blinking and looking at `blinky`

Reviewer 2: Looking at blinky?

> The LED should be blinking and looking at `blinky`

Reviewer 2: ```suggestion
The LED should be blinking as we are now running this binary:
```

work. The missing ones probably rely on interrupts, or some other thing we
haven't ported to work in RISC-V mode yet.

Here is an example using the `blinky` example in RISC-V mode:
```console
$ cargo brr --bin blinky
    Finished `release` profile [optimized] target(s) in 0.06s

> $ cargo rrr --bin blinky

Reviewer: We already had examples of using aliases above.
I think keeping the non-aliased build commands here would make sense - keeps it obvious for folks not using these aliases.
